### ISSUE TITLE
[BUG] Content Security Policy: The page’s settings blocked the loading of a resource at inline (“script-src”).

### ISSUE THREAD

**Issue by andZibralabs** [vitalik/django-ninja#693]
**Describe the bug**
At /docs url, there is a CSP error:
Content Security Policy: The page’s settings blocked the loading of a resource at inline (“script-src”)

Looks like there is some inline script in templates/ninja/swagger.html

**Versions:**
 - Python version: 3.10.8
 - Django version: 4.1.7
 - Django-Ninja version: 0.21.0
 - Pydantic version: 1.10.4

**Comment by vitalik** [comment]
@andZibralabs how to reproduce this ? which browser (browser version) do you use ?

**Comment by andZibralabs** [comment]
@vitalik It is not about the browser, but about the Content-Security-Policy that has been set on the web server.
But the browser I was using is:
Mozilla Firefox 110.0.1
How to reproduce:
Set the Content-Security-Policy: "default-src 'none'; script-src 'self'; img-src 'self' data:; style-src 'self';"
go to url /docs

**Comment by vitalik** [comment]
> Set the Content-Security-Policy

so this is not default browser behaviour? you just enforce it manually ?

I do not see any error messages:
<img width="1413" alt="SCR-20230309-opa" src="https://user-images.githubusercontent.com/95222/224077457-e4fa0e9d-4e2c-41ed-875f-051bae9f78ab.png">

**Comment by tykling** [comment]
hello @vitalik :)

You can read about the `Content-Security-Policy` [here](https://content-security-policy.com/ )

It is an HTTP header you add on the webserver to instruct the webbrowser where to accept different types of content from. For example you can make a CSP header telling the browser to only accept images from `cdn.example.com` and js from `scripts.example.com` - making XSS attacks impossible because the browser now knows not to load javascript from `evilhacker.example.org` even if an XSS vuln should exist.

The problem with inlining scripts like the current state before #695 is that to the browser it looks exactly like a script injection. The only way to make it work with CSP is to effectively water down the CSP header so much that you might as well not have it.

TL;DR: Never use inline <script> tags, always load JS from external files. The same for images, don't use data: in <img> tags, load external images instead. It helps the browser keep the users safe.

**Comment by tykling** [comment]
Hello again :)

Is there anything I can to to help get #695 merged? At the moment we have to set an unsafe CSP header on the webservers where we use django-ninja API docs. This is a shame since the fix is simple.

Thanks!

**Comment by fretchen** [comment]
Here, is a way to work with a CSP that seemed to work for me.

Let's assume that you are modifying the [swagger.html](https://github.com/vitalik/django-ninja/blob/master/ninja/templates/ninja/swagger.html) file on your own. Than it would seem that the critical part is:

```html
script>
        const ui = SwaggerUIBundle({
            url: '{{ openapi_json_url }}',
            dom_id: '#swagger-ui',
            presets: [
            SwaggerUIBundle.presets.apis,
            SwaggerUIBundle.SwaggerUIStandalonePreset
            ],
            layout: "BaseLayout",
        {% if api.csrf and csrf_token %}
            requestInterceptor: (req) => {
				req.headers['X-CSRFToken'] = "{{csrf_token}}"
				return req;
			},
        {% endif %}
            deepLinking: true
        })
    </script>
```

Following the documentation of django-csp this can be made safe through an inline nonce 

1. Loading the csp via `{% load csp %}` just below `{% load static %}`
2.  Wrapping the js code as such:
```
{% script type="application/javascript" async=False %}
<script>
    const ui = SwaggerUIBundle({
        url: '{{ openapi_json_url }}',
        dom_id: '#swagger-ui',
        presets: [
            SwaggerUIBundle.presets.apis,
            SwaggerUIBundle.SwaggerUIStandalonePreset
        ],
        layout: "BaseLayout",
        {% if api.csrf and csrf_token %}
    requestInterceptor: (req) => {
        req.headers['X-CSRFToken'] = "{{csrf_token}}"
        return req;
    },
        {% endif %}
    deepLinking: true
        })
</script>
{% endscript %}
```

This then needs the extension of the settings to include:

```
CSP_INCLUDE_NONCE_IN = {
    "script-src": ["<nonce>"],
}
```

All of this becomes unnecessary by #695 from my understanding.

**Comment by vitalik** [comment]
released in 0.22

**Comment by CubicrootXYZ** [comment]
Running into the same issue when using `docs=Redoc()` - can we move the inline JS there to a file as well?
